bukkit/src/main/java/us/myles/ViaVersion/bukkit/handlers/BukkitDecodeHandler.java

```diff
@@ -26,58 +23,19 @@ public BukkitDecodeHandler(UserConnection info, ByteToMessageDecoder minecraftDe
 
     @Override
     protected void decode(ChannelHandlerContext ctx, ByteBuf bytebuf, List<Object> list) throws Exception {
-        // use transformers
-        if (bytebuf.readableBytes() > 0) {
-            // Ignore if pending disconnect
-            if (info.isPendingDisconnect()) {
-                return;
-            }
-            // Increment received
-            boolean second = info.incrementReceived();
-            // Check PPS
-            if (second) {
-                if (info.handlePPS())
-                    return;
-            }
+        ByteBuf draft = ctx.alloc().buffer().writeBytes(bytebuf);
```

Comment:
Wasted CPU by copying on native versions?

Reply:
You'd be correct. I actually have a version of `BukkitEncodeHandler` that avoids copying altogether, and we've been running it in production without problems.

Reply:
@astei Would you consider contributing this upstream? As this is definitely very beneficial.

Reply:
I would be happy to bring this upstream. I think there are opportunities for more aggressive optimizations we can investigate.

Reply:
Here is the `BukkitEncodeHandler` from Mineteria: https://gist.github.com/astei/38d40d4769398ee7d189764240a12ee2

As the original author of the code, I hereby license it under the same terms as ViaVersion itself (the MIT license).

Features:
* Avoids memory copies where possible
* Avoids encoding if client and server versions match

Reply:
AFAIK when the client version and the server version are the same user.isActive() should be false. Is there any reason for the extra check?

Reply:
Seems not. The original code did correctly avoid a memory copy if the server and client versions matched. I just wrote this out explicitly as it's cleaner.